File: app/src/main/java/com/kamelong/aodia/EditTrainType/EditTrainTypeFragment.java
package com.kamelong.aodia.EditTrainType;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;

import androidx.annotation.NonNull;

import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.kamelong.aodia.AOdia;
import com.kamelong.OuDia.LineFile;
import com.kamelong.OuDia.TrainType;
import com.kamelong.aodia.AOdiaFragmentCustom;
import com.kamelong.aodia.R;
import com.kamelong.tool.SDlog;

import java.util.ArrayList;

public class EditTrainTypeFragment extends AOdiaFragmentCustom {
    int fileIndex;
    LineFile lineFile;
    View fragmentContainer;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        try {//まずBundleを確認し、fileNum,diaNumber,directを更新する
            Bundle bundle = getArguments();
            fileIndex = bundle.getInt(AOdia.FILE_INDEX, 0);
        } catch (Exception e) {
            SDlog.log(e);
        }
        //Fragmentのレイアウトxmlファイルを指定し、メインのViewをfragmentContainerに代入する（つまり消すな）
        fragmentContainer = inflater.inflate(R.layout.traintype_edit_fragment, container, false);
        try {
            lineFile = getAOdia().getLineFile(fileIndex);
        } catch (Exception e) {
            SDlog.log(e);
        }
        return fragmentContainer;
    }

    private void init() {
        if (lineFile == null) {
            getAOdia().killFragment(this);
            return;
        }
        final LinearLayout typeListLinear = fragmentContainer.findViewById(R.id.typeListLinear);
        typeListLinear.removeAllViews();
        for (int i = 0; i < lineFile.trainType.size(); i++) {
            typeListLinear.addView(new EditTrainTypeView(getMainActivity(), lineFile.trainType.get(i)));
        }

    }


    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        if (lineFile == null) {
            getAOdia().killFragment(this);
            return;
        }

    }
    @Override
    public void onStart(){
        super.onStart();
        init();
        final FloatingActionButton copyButton = fragmentContainer.findViewById(R.id.copyButton);
        copyButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                LinearLayout typeListLinear = fragmentContainer.findViewById(R.id.typeListLinear);
                ArrayList<TrainType> copyTrainType = new ArrayList<>();
                for (int i = 0; i < lineFile.trainType.size(); i++) {
                    if (((EditTrainTypeView) typeListLinear.getChildAt(i)).checked) {
                        copyTrainType.add(lineFile.getTrainType().get(i));
                    }
                }
                getAOdia().copyTrainType = copyTrainType;
                SDlog.toast("列車種別をコピーしました");
            }
        });
        final FloatingActionButton deleteButton = fragmentContainer.findViewById(R.id.deleteButton);
        deleteButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                LinearLayout typeListLinear = fragmentContainer.findViewById(R.id.typeListLinear);
                ArrayList<TrainType> deleteType = new ArrayList<>();
                for (int i = 0; i < lineFile.trainType.size(); i++) {
                    if (((EditTrainTypeView) typeListLinear.getChildAt(i)).checked) {
                        deleteType.add(lineFile.getTrainType().get(i));
                    }
                }
                for (TrainType type : deleteType) {
                    if (lineFile.deleteTrainType(type)) {

                    } else {
                        SDlog.toast(type.name + ":路線内で使われているため、削除できません");
                        break;
                    }
                }
                init();
            }
        });
        final FloatingActionButton pasteButton = fragmentContainer.findViewById(R.id.pasteButton);
        pasteButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                LinearLayout typeListLinear = fragmentContainer.findViewById(R.id.typeListLinear);
                ArrayList<TrainType> copyTrainType = getAOdia().copyTrainType;
                int i = 0;
                for (i = 0; i < lineFile.trainType.size(); i++) {
                    if (((EditTrainTypeView) typeListLinear.getChildAt(i)).checked) {
                        break;
                    }
                }
                for (TrainType type : copyTrainType) {
                    lineFile.addTrainType(i, type.clone());
                    i++;
                }
                SDlog.toast("列車種別を貼り付けました");
                init();
            }
        });
        final FloatingActionButton addButton = fragmentContainer.findViewById(R.id.addButton);
        addButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                LinearLayout typeListLinear = fragmentContainer.findViewById(R.id.typeListLinear);
                ArrayList<TrainType> copyTrainType = getAOdia().copyTrainType;
                int i = 0;
                for (i = 0; i < lineFile.trainType.size(); i++) {
                    if (((EditTrainTypeView) typeListLinear.getChildAt(i)).checked) {
                        break;
                    }
                }
                lineFile.addTrainType(i, new TrainType());
                init();
            }
        });
    }


    @NonNull
    @Override
    public String getName() {
        try {
            String line = lineFile.name;
            if (line.length() > 10) {
                line = line.substring(0, 10);
            }
            return line + "\n" + "列車種別";
        } catch (Exception e) {
            return "列車種別";
        }
    }

    @Override
    public LineFile getLineFile() {
        return lineFile;
    }
}

File: app/src/main/java/com/kamelong/aodia/editTrainType/EditTrainTypeFragment.kt
package com.kamelong.aodia.editTrainType

import android.app.Fragment
import android.os.Bundle
import android.view.KeyEvent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.LinearLayout
import com.kamelong.aodia.AOdiaActivity
import com.kamelong.aodia.AOdiaFragment
import com.kamelong.aodia.AOdiaFragmentInterface
import com.kamelong.aodia.R
import com.kamelong.aodia.diadata.AOdiaDiaFile
import com.kamelong.aodia.editStation.CopyPasteInsertAddDeleteDialog

class EditTrainTypeFragment :Fragment(),AOdiaFragmentInterface{
    override var fragment=this as Fragment
    override lateinit var diaFile: AOdiaDiaFile
    override lateinit var aodiaActivity: AOdiaActivity
    lateinit var fragmentContainer: View
    lateinit var typeListLinear:LinearLayout
    var fileIndex=0

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {

        super.onCreateView(inflater, container, savedInstanceState)
        try {
            aodiaActivity = getActivity() as AOdiaActivity
            fragment = this
            val bundle = arguments
            fileIndex = bundle.getInt("fileNum")
            diaFile = aodiaActivity.diaFiles[fileIndex]
        } catch (e: Exception) {
            e.printStackTrace()
            //activity.killFragment(this)
        }
        fragmentContainer = inflater.inflate(R.layout.edit_traintype_fragment, container, false)
        fragmentContainer.setFocusableInTouchMode(true)
        fragmentContainer.requestFocus();
        fragmentContainer.setOnKeyListener(object : View.OnKeyListener {
            override fun onKey(v: View, keyCode: Int, event: KeyEvent): Boolean {
                println("key")

                return if (keyCode == KeyEvent.KEYCODE_BACK && event.getAction() == KeyEvent.ACTION_UP) {
                    val frameLayout = fragmentContainer.findViewById<FrameLayout>(R.id.frameLayout)
                    true
                } else false
            }
        })
        return fragmentContainer
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        typeListLinear = fragmentContainer.findViewById(R.id.typeListLinear)
        for (i in 0 until diaFile.trainTypeNum) {
            typeListLinear.addView(EditTrainTypeView(activity,diaFile.getTrainType(i)))

        }

        fragmentContainer.setOnLongClickListener {
            true
        }
    }

}